base/asyncmap.jl

```diff
@@ -19,44 +19,97 @@ type AsyncCollector
     f
     results
     enumerator::Enumerate
-    ntasks::Int
+    max_tasks::Function
+    c::Channel{Tuple{Int, Any}}     # to communicate with the tasks
+
+    AsyncCollector(f, r, en::Enumerate, mt::Function, c::Channel) = new(f, r, en, mt, c)
 end
 
 function AsyncCollector(f, results, c...; ntasks=0)
     if ntasks == 0
         ntasks = max(nworkers(), 100)
+        max_tasks = ()->ntasks
+    elseif isa(ntasks, Integer)
+        max_tasks = ()->ntasks
+    elseif isa(ntasks, Function)
+        max_tasks = ntasks
+    else
+        throw(ArgumentError("ntasks must be an Integer or a zero-arg function returning the maximum number of tasks allowed."))
```

Comment:
A `WorkerPool` allows for workers to be dynamically added. `pmap` now passes this keyword arg as `ntasks=()->nworkers(p)` where `p` is the worker pool.